### PR TITLE
Data spill and checkpoint

### PR DESCRIPTION
**Pull request by davidHysom**
This PR adds functionality for "spilling" samples to a local directory, and for checkpointing. "Spilling" means that, when set_conduit_node() is called the conduit node is saved to disk, instead of being saved in RAM (in the m_data data structure). 

Current status: code has been tested using alexnet and a small portion of imagenet for both checkpointing and spilling; code appears to work correctly.

### CODE REVIEW

File: src/data_readers/data_reader.cpp [hunk 1]
```diff
@@ -724,6 +725,7 @@ void generic_data_reader::instantiate_data_store(const std::vector<int>& local_l
       m_data_store->build_preloaded_owner_map(local_list_sizes);
     }
     preload_data_store();
+    m_data_store->set_is_preloaded();
```
[the remark this answers is not in the document]
Comment (benson31) [reply src/data_readers/data_reader.cpp]: If something must be called in every derived class, the Template Method might be a useful implementation strategy.

[the remark this answers is not in the document]
Comment (benson31) [reply src/data_readers/data_reader.cpp]: In particular, if you have an algorithm in which there is `common_behavior()` and `custom_derived_behavior()`, you can implement this in this way:
```
class base
{
public:
  void my_algorithm()
  {
    common_behavior();
    custom_derived_behavior();
  }

private:
  void common_behavior() { /* whatever */ }
  virtual void custom_derived_behavior() = 0;
};// class base

class derived : public base
{
  void custom_derived_behavior() override { /* whatever */ }
};// class derived
```
This has many advantages, not least of which is that the contract of the function is enforced in the base class and all the derived class has to do to "get it right" is implement the portion of the behavior that changes. Basically, when someone implements a derived class, they don't have to worry about going through a checklist of assumptions because they're already enforced. This eases both new development and maintenance.

**Comment by davidHysom** [comment]
Tom,
I don't fully understand your suggestion or how it applies to set_is_preloaded(),
which merely sets a bool. Also, if there's a derived class, how does the data_store
get instantiated? Is it sometimes the base class and sometimes the derived class?

I'd like to talk this over, if you have time available tomorrow before 1pm. Else sometime
next week or the week after -- possibly over the phone -- depending on how my
recovery from surgery turns out.

thanks!
________________________________

**Comment by davidHysom** [comment]
> This seems fine. `data_store_conduit` has unnecessarily `protected` data  ...
I find the print statements very useful for code development and profiling. I always want to know what the code is doing, and how long it takes. That said, the print statements aren't necessarily intended to be in place long-term. In a future PR I'll make them optional, and printed to a file.

**Comment by benson31** [comment]
@davidHysom Basically what I was suggesting is that the last line of `generic_data_reader::preload_data_store()` should be `m_data_store->set_is_preloaded();`. This is a `generic_data_reader` problem, not a `data_store_*` problem.

Suppose that we actually move `m_data_store->set_is_preloaded();` into `preload_data_store()`. The design problem here is: Given that `generic_data_reader::preload_data_store()` is (effectively pure) virtual, how do we make sure that every derived class actually calls this method? The answer is the Template Method:
```cpp
class generic_data_reader
{
public:
  ...
  void preload_data_store()
  {
    do_preload_data_store();
    m_data_store->set_is_preloaded();
  }
  ...
private:
  virtual void do_preload_data_store()
  {
    LBANN_ERROR("Not implemented.");
  }
};
```
Now, all we need to do is change the derived classes so that their current implementations of `preload_data_store` become the implementations of `do_preload_data_store()`:
```cpp
class data_reader_jag_conduit : public generic_data_reader // for example
{
  ...
private: // or protected, if further derived classes need to call the implementation
  void do_preload_data_store() override
  {
    /* Current implementation of data_reader_jag_conduit::preload_data_store().
       This is equivalent to simply changing the name, obviously.
    */
  }
  ...
};
```
The base class implements all common parts of the algorithm, leaving virtual placeholders for the customizable behavior (hence the name "Template Method"). This ensures that "things derived classes must do" are actually done, because they're actually invariants of the base class.

Does that make sense? On the one hand, this change might be outside the scope of this PR. On the other hand, there are only three data readers that implement `preload_data_store()`, so it's not a big diff to push this through now.

File: include/lbann/data_store/data_store_conduit.hpp [hunk 2]
```diff
@@ -198,16 +262,14 @@ protected :
   /// convenience handle
   const std::vector<int> *m_shuffled_indices;
 
-  void exchange_data_by_sample(size_t current_pos, size_t mb_size);
+  /// contains the Nodes that this processor owns;
+  /// maps data_id to conduit::Node
+  mutable std::unordered_map<int, conduit::Node> m_data;
```
[the remark this answers is not in the document]
Comment (davidHysom) [reply include/lbann/data_store/data_store_conduit.hpp]: Well, yes. But  shuffled_indices is std::vector<int>, which is why I used int instead of size_t.

File: src/data_readers/data_reader.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidHysom) [reply src/data_readers/data_reader.cpp]: > It makes more sense to me if we call `set_is_preloaded` inside of `preload_data_store`, even if it requires repeating that function call for every derived class. 
set_is_preloaded was formerly called set_preload. I changed the name to make it (what I think) is more informative. Anyway, set_preload *was* formerly called in the derived data_reader preload_data_store classes. But since preload_data_store is only called in generic_data_reader, it made sense (again, to me) to put the call there.

If there is a use case where preload_data_store() might be called anyplace other than generice_data_reader::instantiate_data_store(), then I agree with you.

File: src/data_store/data_store_conduit.cpp [hunk 3]
```diff
@@ -317,34 +326,43 @@ void data_store_conduit::set_conduit_node(int data_id, conduit::Node &node, bool
...
+      m_spilled_nodes[data_id] = m_cur_spill_dir_integer;
+    }
+
+    else {
+      m_mutex.lock();
+      build_node_for_sending(node, m_data[data_id]);
+      error_check_compacted_node(m_data[data_id], data_id);
+      m_sample_sizes[data_id] = m_data[data_id].total_bytes_compact();
+      m_mutex.unlock();
+    }
```
[the remark this answers is not in the document]
Comment (davidHysom) [reply src/data_store/data_store_conduit.cpp]: > If we might throw an exception while holding a mutex, a RAII approach might be safer:
> 
> If the exception is thrown, then `lock` will go out of scope and release `m_mutex`.

I think your suggested change only guards the call to error_check_compacted_node. Is that true? (note: I'm very much a novice at using mutex). There are three calls that, as I understand it, should be made atomic.

File: src/data_store/data_store_conduit.cpp [hunk 4]
```diff
@@ -1260,4 +1298,289 @@ size_t data_store_conduit::get_num_indices() const {
...
+    m_num_files_in_cur_spill_dir = 0;
+    m_cur_spill_dir_integer += 1;
+    m_cur_spill_dir = get_conduit_dir() + "/" + to_string(m_cur_spill_dir_integer);
+    bool exists = file::directory_exists(m_cur_spill_dir);
+    if (!exists) {
+      file::make_directory(m_cur_spill_dir);
+    }
+  }
+  const std::string fn = m_cur_spill_dir + "/" + std::to_string(data_id);
+  node.save(fn);
```
[the remark this answers is not in the document]
Comment (davidHysom) [reply src/data_store/data_store_conduit.cpp]: Thanks for the catch; you are correct.

File: src/data_readers/data_reader.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (benson31) [reply src/data_readers/data_reader.cpp]: The issue is more proper separation and encapsulation of concerns rather than who calls `preload_data_store`. To me, anyway, it seems that setting the preloaded flag is an integral part of preloading the data store. But as implied by the approval, I'm not going to make a fuss about this if you disagree.

**Comment by davidHysom** [comment]
> how do we make sure that every derived class actually calls this method?

I'll have to think more about your comments tomorrow (on pain killers at the moment). 

What I think is pertinent: derived data_reader classes are not required to implement preload_data_store() -- which is why I didn't make it pure virtual. Of our many data_reader classes, currently only data_reader_jag_conduit,  data_reader_numpy_npz_conduit, and image_data_reader implement the method -- because those are the only data readers that work with the data_store.

There's probably a more sophisticated way to handle this.

**Comment by benson31** [comment]
The point is to avoid sophistication. Sophistication requires effort to decipher, which implies a maintenance burden, which is bad. The Template Method is a well-established design pattern that is meant for this purpose. (I have citations if you're curious.)

I'm aware only three classes implement that method; all others will land on the base class implementation that throws an exception. My proposed change will not affect this behavior at all; the base class's default implementation of `do_preload_data_store()` throws the same exception as before.

Ultimately, though, this is a small nitpick that spiraled off @timmoon10's original comment. My intent was to head off some unnecessary code duplication ("even if it requires repeating that function call for every derived class" <-- this would be a bad move that would cause me to request changes to this PR) and illuminate a nice, oft-underutilized design pattern that can solve these sorts of problems.

File: include/lbann/data_store/data_store_conduit.hpp [hunk 5]
```diff
@@ -144,7 +146,68 @@ class data_store_conduit {
 
   void flush_debug_file(); 
 
+  /// write m_data to file; may also write other data structures
+  /// (e.g, src/data_store/data_store_conduit.cpp) to file
```
Comment (ndryden) [inline include/lbann/data_store/data_store_conduit.hpp]: Referencing a source file is not helpful.

File: include/lbann/data_store/data_store_conduit.hpp [hunk 6]
```diff
@@ -144,7 +146,68 @@ class data_store_conduit {
...
+  void write_checkpoint(std::string dir_name);
+  
+  /// load m_data from local file; may also write other data structures
+  /// (e.g, src/data_store/data_store_conduit.cpp) to file
```
Comment (ndryden) [inline include/lbann/data_store/data_store_conduit.hpp]: Ditto.

File: include/lbann/data_store/data_store_conduit.hpp [hunk 7]
```diff
@@ -144,7 +146,68 @@ class data_store_conduit {
...
+   * "/conduit_<m_rank_in_world>" to the string.
+   */
+  std::string m_spill_dir_base;
+
+  /** @brief Used to form the directory path for spilling conduit nodes */
+  int m_cur_spill_dir_integer = -1;
+
+  /** @brief @brief Current directory for spilling (writing to file) conduit nodes 
+   *
+   * m_cur_dir = m_spill_dir_base/<m_cur_spill_dir_integer>
```
Comment (ndryden) [inline include/lbann/data_store/data_store_conduit.hpp]: Is this supposed to be `m_cur_spill_dir`?

File: include/lbann/data_store/data_store_conduit.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ndryden) [reply include/lbann/data_store/data_store_conduit.hpp]: I don't follow this reasoning?

But perhaps we should take a separate PR to take care of this.

File: include/lbann/data_store/data_store_conduit.hpp [hunk 8]
```diff
@@ -37,6 +37,10 @@
 #include <unordered_set>
 #include <mutex>
 
+#include <cereal/types/unordered_map.hpp>
+#include <cereal/archives/binary.hpp>
```
Comment (ndryden) [inline include/lbann/data_store/data_store_conduit.hpp]: Do we need to include these here?

File: src/data_readers/data_reader.cpp [hunk 9]
```diff
@@ -703,6 +703,7 @@ void generic_data_reader::instantiate_data_store(const std::vector<int>& local_l
   if (m_shuffled_indices.size() == 0) {
     LBANN_ERROR("shuffled_indices.size() == 0");
   }
+  //m_data_store->set_preload();
```
Comment (ndryden) [inline src/data_readers/data_reader.cpp]: Can we delete dead code?

File: src/data_store/data_store_conduit.cpp [hunk 10]
```diff
@@ -55,6 +60,7 @@ data_store_conduit::data_store_conduit(
   m_world_master = m_comm->am_world_master();
   m_trainer_master = m_comm->am_trainer_master();
   m_rank_in_trainer = m_comm->get_rank_in_trainer();
+  m_rank_in_world = m_comm->get_rank_in_world();
   m_np_in_trainer = m_comm->get_procs_per_trainer();
```
Comment (ndryden) [inline src/data_store/data_store_conduit.cpp]: Why are we duplicating information?

(This should probably be addressed in a separate PR.)

File: src/data_store/data_store_conduit.cpp [hunk 11]
```diff
@@ -296,17 +301,21 @@ void data_store_conduit::set_conduit_node(int data_id, conduit::Node &node, bool
...
   }
-  //m_mutex.lock();
+
+  m_mutex.lock();
   if (already_have == false && m_data.find(data_id) != m_data.end()) {
+    m_mutex.unlock();
     LBANN_ERROR("duplicate data_id: ", data_id, " in data_store_conduit::set_conduit_node; role: ", m_reader->get_role());
   }
-
+  m_mutex.unlock();
```
Comment (ndryden) [inline src/data_store/data_store_conduit.cpp]: The methods of `std::mutex` should in general not be used directly; it is meant to be used in `std::unique_lock` or `std::lock_guard` (and so on), to provide exception-safe locking with RAII semantics.

File: src/data_store/data_store_conduit.cpp [hunk 12]
```diff
@@ -597,7 +619,7 @@ void data_store_conduit::build_preloaded_owner_map(const std::vector<int>& per_r
 
 #if 0
```
Comment (ndryden) [inline src/data_store/data_store_conduit.cpp]: Dead code should be deleted. If it serves a useful testing purpose, write a unit test. If it's meant for debugging/informational purposes, make it toggleable, since it will probably be more widely useful. Otherwise it can always be retrieved from the Git history if needed.

**Comment by davidHysom** [comment]
>> +  /// contains the Nodes that this processor owns;
>+  /// maps data_id to conduit::Node
>+  mutable std::unordered_map<int, conduit::Node> m_data;
>I don't follow this reasoning?

I think you're asking "why is this mutable?" You should ask Brian.
OK, seriously, in copy_members we have:
  rhs.m_data.erase(i);
And the rhs may come from the copy ctor, so rhs must be const.

________________________________

**Comment by davidHysom** [comment]
> Basically what I was suggesting

OK, this makes sense.

File: src/data_readers/data_reader.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidHysom) [reply src/data_readers/data_reader.cpp]: I will make these changes in a follow on PR.

File: src/data_store/data_store_conduit.cpp (hunk 10 above)
Comment (davidHysom) [reply src/data_store/data_store_conduit.cpp]: will adress in a separate PR: too many places to make changes, and not relevant to this PR's purpose.

File: include/lbann/data_store/data_store_conduit.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidHysom) [reply include/lbann/data_store/data_store_conduit.hpp]: My less-than and greater-than brackets got squashed. Meant to say that shuffled_indices is currently a vector of integers. So it doesn't make sense to change m_data unless we also change shuffled_indices, and who knows whee else that path will lead. This is work for a future PR.

File: src/data_store/data_store_conduit.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidHysom) [reply src/data_store/data_store_conduit.cpp]: modified.

[the remark this answers is not in the document]
Comment (timmoon10) [reply src/data_store/data_store_conduit.cpp]: `std::lock_guard` calls `m_mutex.lock()` in the constructor and `m_mutex.unlock()` in its destructor, so the lock will release the mutex when it reaches the end of the `else` statement or if an exception is thrown.
